wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/ProtoFile.kt

```diff
@@ -129,24 +129,40 @@ data class ProtoFile(
     return result
   }
 
+  private fun referencedTypes(schema: Schema): List<Type> {
+    val types = typesAndNestedTypes()
+    val messages = types.filterIsInstance<MessageType>()
+
+    val messageTypes = messages
+      .flatMap { it.fieldsAndOneOfFields }
+      .mapNotNull { it.type }
+
+    val messageOptions = messages.map { it.options }
+    val fieldOptions = messages.flatMap { it.fieldsAndOneOfFields.map { it.options } }
+    val oneOfOptions = messages.flatMap { it.oneOfs }.map { it.options }
+    val optionTypes = (messageOptions + oneOfOptions + fieldOptions).flatMap { it.fields().asMap().keys }
+
+    val rpcTypes = services
+      .flatMap { service -> service.rpcs }
+      .flatMap { rpc -> listOfNotNull(rpc.requestType, rpc.responseType) }
+
+    return (messageTypes + optionTypes + rpcTypes).distinct().mapNotNull { schema.getType(it) }.toList()
```

Comment:
* `messageTypes` are the field types defined in messages
* `optionTypes` are the option types for the message and fields, and one-ofs
  * The import paths of any option extensions are resolved below in `retainImports`
* `rpcTypes` are the request and response types defined in services

Reply:
Not sure if we need to retain enum's and constants' options?

Reply:
I remembered later that I found a list of a bunch of different option types somewhere. I’ll take a look at that and see if there’s others to add as well.

Reply:
I've update this now to look for all of [these option types](https://github.com/square/wire/blob/master/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt#L405-L419) (at least where I believe they would be in the model). Please confirm the new logic makes sense.